### ISSUE TITLE
win32: can read process memory using seekdir/readdir 

### ISSUE THREAD

**Issue by AlexDavies501** [Perl/perl5#20768]
This applies to win32 Perl.

There is no bounds checking on the POS argument passed to `seekdir(DIRHANDLE, POS)`.
So the subsequent `readdir(DIRHANDLE)` will read from arbitrary memory - or SEGV.

This allows for reading the memory space:

```
use warnings;
use strict;
$| = 1;

opendir my $dh, '.' or die "! can not opendir('.'): $!\n";
my @f = readdir $dh;

my $pos = 0;

# Keep going until SEGV
while (1) {
	seekdir $dh, $pos;
	my $bytes = readdir $dh;
	if (defined $bytes) {
		my $len = length $bytes;
		if ($len > 5) { # do not print short strings
			$bytes =~ s|[^[:print:]]+||g;
			print "$pos => $bytes\n";
		}
		$pos += $len;
	}
	++$pos; # skip NUL
}
```

The above prints out strings from the memory space until it SEGVs.
(This incidentally shows perl storing a lot of copies of the path to "warnings.pm"!)

Suggested (and tested) patch fix:

```
--- a/win32/win32.c
+++ b/win32/win32.c
@@ -1016,7 +1016,12 @@ win32_telldir(DIR *dirp)
 DllExport void
 win32_seekdir(DIR *dirp, long loc)
 {
-    dirp->curr = loc == -1 ? NULL : dirp->start + loc;
+    /* Ensure dirp->curr remains within `dirp->start` buffer. */
+    if (loc >= 0 && dirp->end - dirp->start > (ptrdiff_t) loc) {
+        dirp->curr = dirp->start + loc;
+    } else {
+        dirp->curr = NULL;
+    }
 }

 /* Rewinddir resets the string pointer to the start */
```

**Comment by demerphq** [comment]
If at all possible we prefer a PR so we can add you to the AUTHORS list correctly. If you prefer not to be listed one of us can prepare the PR from your patch. If you make a PR after you commit your change run Porting/updateAUTHORS.pl and review and commit the changes as a second commit and then create the PR.

**Comment by AlexDavies501** [comment]
Have created a PR with a single commit: https://github.com/Perl/perl5/pull/20792

Contains changes to AUTHORS/.mailmap for my old email addresses.

**Comment by demerphq** [comment]
Thanks. I just kicked off the CI. Do you think you can rework your original test case here into a test that we can use to verify this is working. BTW, is (This incidentally shows perl storing a lot of copies of the path to "warnings.pm"!) still the case on blead perl (before this patch is applied?) I would expect not. (I reworked how that data is stored and i think there should be far less copies, maybe even just one.)

**Comment by tonycoz** [comment]
Fixed by #20792
